### PR TITLE
Allow operations to opt out of auth middleware

### LINKED ISSUE #99: Allow operations (actions and queries) to opt out of `auth` middleware on server
Right now, if `auth` feature is used, all actions and queries have `auth` middleware in front of them.

But, this is problematic in some situations, for example if you have `signUp` action that calls `createNewUser`, it will get rejected with 401 if JWT token is wrong. Which makes sense, but we actually want `signUp` to succed in situation like this, because how are we going to fix that JWT token otherwise? So, what we really want is for `signUp` to not require authentication, which makes sense, but, we don't have that right now as an option.

This should probably work smth like this:
```
action signUp {
  fn: import { signUp } from "@ext/actions.js",
  entities: [User],
  auth: false
}
```
and in this case `auth` middleware should not be applied on server on this action.

I am marking this as enhancement because that is what it is, but also as bug because it actually limits user severly in situation like this and manifests as kind of a bug.

Draft:
1. Update Wasp/Query.hs and Wasp/Action.hs to have `auth :: Bool` field.
2. Update Parser/Operation.hs to correctly parse the `auth` property.
3. In ServerGenerator, update operations generation to not put `auth` middleware on operations that have `auth` specified as `False`.

### PR DESCRIPTION
**Pull request by saatvikshah**
# Description

This patch allows operations to individually specify whether they want
to use or opt out of auth middleware, via an additional `auth` argument
under the actions/queries sections. If left unspecified it is assumed to
be true.

This is addressed by removing the current global auth middleware
registration, and adding it instead to individual operation route
specification.

Fixes #99 

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: waspc/data/Generator/templates/server/src/routes/operations/index.js [hunk 1]
```diff
@@ -11,12 +11,8 @@ import {= importIdentifier =} from '{= importPath =}'
 
 const router = express.Router()
 
-{=# isAuthEnabled =}
-router.use(auth)
-{=/ isAuthEnabled =}
-
 {=# operationRoutes =}
-router.post('{= routePath =}', {= importIdentifier =})
+router.post('{= routePath =}', {=# isAuthEnabled =} auth, {=/ isAuthEnabled =} {= importIdentifier =})
```
Comment (saatvikshah) [inline waspc/data/Generator/templates/server/src/routes/operations/index.js:15]: Never really worked with ExpressJS routers but a basic readup showed that adding the `auth` as an argument to `post/get` and before the callback should also install it as middleware.

**Review by Martinsos — COMMENTED** [review]
@saatvikshah wow this looks very good! I think this is it, I left some comments but the general solution is there.

Let's take care of the comments and I think we are good to go!

Did you test this on any of the example apps, maybe waspc/examples/todoApp? If not, we should do that also. It would be best if we add a query that has `auth: false` and use it somehow in the app, to show that it works when user is not authenticated. Hmmm. As described in #99, this should be an operation that is run when user isn't logged in at all -> then `auth: false` is important and enables the operation to work. So I imagine it is a query that shows some info on the login/signup page. It could be a query that returns number of tasks all together in the database. So you know, under the login/signup form, there would be "N tasks created with TodoApp so far!", kind of like a teaser for the person to join the app because look, so many others have created tasks with it. And that would be a super simple query that has `auth` set to false, otherwise it will not work for users that are not logged in.

How does this sound, is this something you could add as a part of this PR?

**Comment by saatvikshah** [comment]
Yep sure, sounds good - I can try to set this up as part of this PR.

**Comment by Martinsos** [comment]
> Yep sure, sounds good - I can try to set this up as part of this PR.

Awesome, ping me if you need assistance with anything!

**Comment by Martinsos** [comment]
Hey @saatvikshah , how is it going, do you need any help with finalizing this?

**Comment by saatvikshah** [comment]
Hi @Martinsos, apologies have recently moved to a new city and was getting settled in leading to the delay. I'll try to have something out in the next couple of days. (the comments are addressed, I need to figure out how to add the view counter part to the todo app)

**Comment by Martinsos** [comment]
@saatvikshah no worries, I just wanted to check if all is ok :)! And congrats on the move that must be exciting, I hope it went well!

**Comment by saatvikshah** [comment]
@Martinsos Thanks! Yea its definitely an exciting change, but comes with the burden of house hunting and paper work :)
I've spent a few hours on this, but am indeed a little stuck in trying to add the task counter to the Todo app. Any ideas on what I might be missing - it seems to be failing with `Attempted import error: 'getNumTasks' is not exported from '../../../queries/getNumTasks.js'.`, however I can clearly see the js file generated and exporting this function.

Also, was the video from your wasp tutorial still available online?

File: waspc/examples/todoApp/ext/pages/auth/Login.js [hunk 2]
```diff
@@ -1,16 +1,24 @@
 import React, { useState } from 'react'
 import { Link } from 'react-router-dom'
+import { useQuery } from '@wasp/queries'
 
 import LoginForm from '@wasp/auth/forms/Login'
+import { getNumTasks } from '@wasp/queries/getNumTasks.js'
```
Comment (Martinsos) [inline waspc/examples/todoApp/ext/pages/auth/Login.js:6]: This seems to be the problem -> you need to use default import!
So it should be

```js
import getNumTasks from '@wasp/queries/getNumTasks'
```

That is just how the Wasp generates the code.

I can see thought that this can be confusing, since you defined your function as named export and you could have assumed that it would be named export again. In the future we will probably work around making this more intuitive, so that you will import the function very similarly, or even the same, as is the path to the file where you defined it.

Comment (Martinsos) [reply waspc/examples/todoApp/ext/pages/auth/Login.js:6]: @saatvikshah

**Comment by Martinsos** [comment]
> @Martinsos Thanks! Yea its definitely an exciting change, but comes with the burden of house hunting and paper work :)
> I've spent a few hours on this, but am indeed a little stuck in trying to add the task counter to the Todo app. Any ideas on what I might be missing - it seems to be failing with `Attempted import error: 'getNumTasks' is not exported from '../../../queries/getNumTasks.js'.`, however I can clearly see the js file generated and exporting this function.
> 
> Also, was the video from your wasp tutorial still available online?

Yes, the video just became available, it is right here: https://www.youtube.com/watch?v=KBoZ7AfEOzY !

**Comment by Martinsos** [comment]
@saatvikshah I just realized one thing -> there could be a case where user does not write `auth` declaration in `main.wasp`, but does provide `auth` fields for operations. In that case, we should throw an error that `auth` is not enabled so they can't specify if specific operation uses auth or not, since it is not enabled all together.

There are multiple places where we could do this check -> but to keep it simple, for now you could do it in the generator (OperationsRoutesG.hs) -> just check if there is situation where auth is not enabled but some of the operations has `auth`. If so, throw error with `error`. This is certainly not the ideal way to do it, but we don't yet have a nice way to handle errors coming from the generator so this will work for now. In the future, when we implement proper support for throwing errors from the generator, we can throw error in a nicer way, or even better we can add another phase in Parser that does semantic checks like this one and do this there. Actually that is probably the right way to do it.

Ok, to summarize: would be great if you can do that check in generator right now with `error`, but pls add a TODO which says this should in the future be moved to Analyzer and become a semantic check, since it doesn't depend on the Generator at all. We could even create an issue github for it.

File: waspc/examples/todoApp/ext/queries.js [hunk 3]
```diff
@@ -14,6 +14,12 @@ export const getTasks = async (args, context) => {
   return tasks
 }
 
+export const getNumTasks = async (args, context) => {
+  const Task = context.entities.Task
+  const numTasks = await Task.count()
+  return numTasks
```
Comment (Martinsos) [inline waspc/examples/todoApp/ext/queries.js:20]: We could we wild and put it all into one line: `return context.entities.Task.count()`

File: waspc/examples/todoApp/ext/pages/auth/Login.js [hunk 4]
```diff
@@ -1,16 +1,24 @@
...
     <>
       <LoginForm/>
       <br/>
       <span>
         I don't have an account yet (<Link to="/signup">go to signup</Link>).
       </span>
+      <br/>
+      <span>
+        Number of tasks already created: 
+      </span>
```
Comment (Martinsos) [inline waspc/examples/todoApp/ext/pages/auth/Login.js:21]: We could also put this on signup form? I would put it on both sign up and login form! And if we have to choose between them, then better just signup form, since that is where "new users" will come first and we want to impress them with how many tasks have already been created with Todo App :D!

Comment (saatvikshah) [reply waspc/examples/todoApp/ext/pages/auth/Login.js]: Done, moved to the Signup page

File: waspc/examples/todoApp/ext/queries.js (hunk 3 above)
Comment (saatvikshah) [reply waspc/examples/todoApp/ext/queries.js]: done

**Comment by saatvikshah** [comment]
Addressed the edge case in OperationRoutesG.hs.
But there is still something wrong, because after rebuilding the app I get a `NetworkError` - `xhr.js:177 POST http://localhost:3001/operations/get-num-tasks net::ERR_CONNECTION_REFUSED`. Does the DB connection(prisma) require authentication maybe?

**Comment by Martinsos** [comment]
> Addressed the edge case in OperationRoutesG.hs and fixed the earlier error.
> But there is still something wrong, because after rebuilding the app I get a `NetworkError` - `xhr.js:177 POST http://localhost:3001/operations/get-num-tasks net::ERR_CONNECTION_REFUSED`. Does the DB connection(prisma) require authentication maybe?

Well that is very weird! Never saw that error before hm. I just pulled your code and tried it, and it works fine for me!
I did:
```
wasp clean
wasp db migrate-dev
wasp start
```
and went to sign-up, saw that 0 tasks are listed, created new user (signed up), created a couple of tasks, loged-out, went again to sign up, saw that number of tasks is correctly listed!

Can you try resetting it all like this and see if it works well then?

File: waspc/src/Generator/ServerGenerator/OperationsRoutesG.hs [hunk 5]
```diff
@@ -79,25 +79,34 @@ relPosixPathFromOperationsRoutesDirToSrcDir :: Path Posix (Rel OperationsRoutesD
 relPosixPathFromOperationsRoutesDirToSrcDir = [reldirP|../..|]
 
 genOperationsRouter :: Wasp -> FileDraft
-genOperationsRouter wasp = C.makeTemplateFD tmplFile dstFile (Just tmplData)
+genOperationsRouter wasp
+   -- TODO: Move below check to Analyzer and make a semantic check
+  | usingAuthOnAnyOperation && (not isAuthEnabled) = error "Auth cannot be used on operations if it is disabled!"
```
Comment (Martinsos) [inline waspc/src/Generator/ServerGenerator/OperationsRoutesG.hs:84]:
```suggestion
  | usingAuthOnAnyOperation && (not isAuthEnabled) = error "`auth` cannot be specified for specific operations if it is not enabled for the whole app!"
```

File: waspc/src/Generator/ServerGenerator/OperationsRoutesG.hs [hunk 6]
```diff
@@ -79,25 +79,34 @@ relPosixPathFromOperationsRoutesDirToSrcDir :: Path Posix (Rel OperationsRoutesD
 relPosixPathFromOperationsRoutesDirToSrcDir = [reldirP|../..|]
 
 genOperationsRouter :: Wasp -> FileDraft
-genOperationsRouter wasp = C.makeTemplateFD tmplFile dstFile (Just tmplData)
+genOperationsRouter wasp
+   -- TODO: Move below check to Analyzer and make a semantic check
```
Comment (Martinsos) [inline waspc/src/Generator/ServerGenerator/OperationsRoutesG.hs:83]:
```suggestion
   -- TODO: Right now we are throwing error here, but we should instead perform this check in parsing/analyzer phase, as a semantic check, since we have all the info we need then already.
```

**Review by Martinsos — APPROVED** [review]
@saatvikshah this is looking good to me, and it works for me!
I added two small comments, pls just take care of them and I will merge it then!

File: waspc/src/Generator/ServerGenerator/OperationsRoutesG.hs (hunk 6 above)
Comment (saatvikshah) [reply waspc/src/Generator/ServerGenerator/OperationsRoutesG.hs]: done

File: waspc/src/Generator/ServerGenerator/OperationsRoutesG.hs (hunk 5 above)
Comment (saatvikshah) [reply waspc/src/Generator/ServerGenerator/OperationsRoutesG.hs]: done

**Comment by saatvikshah** [comment]
Thanks, addressed your comments!

Hmm, I'm pretty sure this used to work in the past, but `stack exec wasp db migrate-dev` is now giving me
```
Error: P3015

Could not find the migration file at Documents/wasp/waspc/examples/todoApp/.wasp/out/db/migrations/20210618232118_/migration.sql. Please delete the directory or restore the migration file.
```

It might just be something wrong on my end, so I'll try to setup the wasp project from scratch again locally.

**Comment by Martinsos** [comment]
> Thanks, addressed your comments!
> 
> Hmm, I'm pretty sure this used to work in the past, but `stack exec wasp db migrate-dev` is now giving me
> 
> ```
> Error: P3015
> 
> Could not find the migration file at Documents/wasp/waspc/examples/todoApp/.wasp/out/db/migrations/20210618232118_/migration.sql. Please delete the directory or restore the migration file.
> ```
> 
> It might just be something wrong on my end, so I'll try to setup the wasp project from scratch again locally.

Hmmmm, maybe running `wasp start` before that `wasp db migrate-dev` would help. We still have some kinks here that need to be taken care of, so we sometimes get in problems like this.

Cool! I will treat this PR as done :). The only thing missing is: resolving conflicts and making sure that code-formatter passes (you can see the check fail above). I will take a look if I can take care of those right now.

**Comment by Martinsos** [comment]
> > Thanks, addressed your comments!
> > Hmm, I'm pretty sure this used to work in the past, but `stack exec wasp db migrate-dev` is now giving me
> > ```
> > Error: P3015
> > 
> > Could not find the migration file at Documents/wasp/waspc/examples/todoApp/.wasp/out/db/migrations/20210618232118_/migration.sql. Please delete the directory or restore the migration file.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It might just be something wrong on my end, so I'll try to setup the wasp project from scratch again locally.
> 
> Hmmmm, maybe running `wasp start` before that `wasp db migrate-dev` would help. We still have some kinks here that need to be taken care of, so we sometimes get in problems like this.
> 
> Cool! I will treat this PR as done :). The only thing missing is: resolving conflicts and making sure that code-formatter passes (you can see the check fail above). I will take a look if I can take care of those right now.

Actually there are no conflicts, Github gave a message that there are conflicts and it can't merge because it wanted to use rebase strategy and this branch has couple of merges happening on it.

**Comment by saatvikshah** [comment]
Thanks for guiding me through this! Happy to continue contributing if you could point me to specific first issues you think would suit me.

Also is the Ormolu formatting documented anywhere?

**Comment by Martinsos** [comment]
> Thanks for guiding me through this! Happy to continue contributing if you could point me to specific first issues you think would suit me.
> 
> Also is the Ormolu formatting documented anywhere?

No problem, it was my pleasure, you did a great job for the first PR, very independent!

Sure, we would love to have you contribute some more -> is there anything specific you are interested in? More Haskell, more JS, is mix ok? Refactoring, or adding new features? Some specific part of Wasp that excites you? More difficult than this issue or was this the right difficulty?

Ormolu -> yes, we have some info! It was very scarce though, I enriched it a bit now: https://github.com/wasp-lang/wasp/tree/master/waspc#formatting .

**Comment by saatvikshah** [comment]
> More Haskell, more JS, is mix ok?
Mix is ok, just as a fyi I have some basic Haskell background and I'm trying to improve at it, while being a beginner at JS which Im trying to learn (day job is C++ 😅)

> Refactoring, or adding new features?
Fine with both, as long as its challenging

>  Some specific part of Wasp that excites you?
I'm generally interested in (1) How to write DSLs and a compiler for them in Haskell (2) Learning a bit more about modern web stacks; in that order. I think its too soon to say which specific part of Wasp I'm most interested in, but anything that touches the guts of the compiler would certainly be fun.

> More difficult than this issue or was this the right difficulty?
A few more issues at this level of difficulty. If you didn't mind fielding more of my questions on the web app side of things then more difficult works too.

**Comment by Martinsos** [comment]
> > More Haskell, more JS, is mix ok?
> 
> Mix is ok, just as a fyi I have some basic Haskell background and I'm trying to improve at it, while being a beginner at JS which Im trying to learn (day job is C++ )
> 
> > Refactoring, or adding new features?
> 
> Fine with both, as long as its challenging
> 
> > Some specific part of Wasp that excites you?
> 
> I'm generally interested in (1) How to write DSLs and a compiler for them in Haskell (2) Learning a bit more about modern web stacks; in that order. I think its too soon to say which specific part of Wasp I'm most interested in, but anything that touches the guts of the compiler would certainly be fun.
> 
> > More difficult than this issue or was this the right difficulty?
> 
> A few more issues at this level of difficulty. If you didn't mind fielding more of my questions on the web app side of things then more difficult works too.

Thanks, this gives me a better idea of what you are interested in!
I created https://github.com/wasp-lang/wasp/issues/293, take a look at that.
Some of these touch the compiler, although they are also often more complex.
There will also be more features regarding the compiler in the upcoming months, as we are building a new version of compiler, and I believe soon we will have a bunch of issues on how to improve / polish / finalize that new version.

In the meantime, if you want to track development of new compiler from scratch, I would advise you take a look at https://github.com/wasp-lang/wasp/pull/274 -> here @craigmc08 wrote new Parser from scratch, together with the skeleton for the other parts that make the "analysis" part of compiler (where we split compiler into "analysis" and "synthesis"(generation)), and after that you can take a look at https://github.com/wasp-lang/wasp/pull/289, which is the next part of "analysis" - type checker. This is pretty complex, but there is a ton of comments so it might be a unique opportunity to see creation of compiler publicly discussed and dissected - and if you have any questions we are happy to answer, it is not a bother, you can ask them on Discord easily.
